### PR TITLE
[ISSUE #86] Add Interceptor for producer and consumer.

### PR DESCRIPTION
**Pull request by xujianhai666**
## What is the purpose of the change

Add interceptor for producer.

## Brief changelog

1. refactor the producer initialization. using option instead of a complete ProducerOption, which is confusing while using the API.
2. add Invoker and Interceptor func type for send operation. The invoke represent a send operation, while the Interceptor will intercept the invoke for logging、metrics or  retry.

### CODE REVIEW

**Comment by dyxushuai** [comment]
Using private config members to protect them when use functional options. [grpc example](https://github.com/grpc/grpc-go/blob/263405c7fe47948606651e13ea1c47b30622de90/dialoptions.go)

**Comment by xujianhai666** [comment]
> Using private config members to protect them when use [functional options](https://halls-of-valhalla.org/beta/articles/functional-options-pattern-in-go,54/). [grpc example](https://github.com/grpc/grpc-go/blob/263405c7fe47948606651e13ea1c47b30622de90/dialoptions.go)

there are some other work to do with that, not included here. we can do that when dealing with https://github.com/apache/rocketmq-client-go/issues/92. If we protect the private config using lower case, then we can not use the method or structure, such as ProducerOptions、DefaultProducerOptions，If we move the code to internal package, we will encounter the "import cycle" problem.

File: internal/consumer/push_consumer.go [hunk 1]
```diff
@@ -99,9 +104,37 @@ func NewPushConsumer(consumerGroup string, opt primitive.ConsumerOption) (PushCo
...
 		p.submitToConsume = p.consumeMessageCurrently
 	}
+
+	ChainInterceptor(p)
+
 	return p, nil
 }
 
+// ChainInterceptor chain list of interceptor as one interceptor
+func ChainInterceptor(p *pushConsumer) {
```
Comment (wenfengwang) [inline internal/consumer/push_consumer.go]: should be unexported?

File: internal/consumer/push_consumer.go [hunk 2]
```diff
@@ -99,9 +104,37 @@ func NewPushConsumer(consumerGroup string, opt primitive.ConsumerOption) (PushCo
...
+	interceptors := p.option.Interceptors
+	switch len(interceptors) {
+	case 0:
+		p.interceptor = nil
+	case 1:
+		p.interceptor = interceptors[0]
+	default:
+		p.interceptor = func(ctx context.Context, req, reply interface{}, invoker primitive.CInvoker) error {
```
Comment (wenfengwang) [inline internal/consumer/push_consumer.go]: use iteration, no recursive

File: internal/consumer/push_consumer.go (hunk 1 above)
Comment (xujianhai666) [reply internal/consumer/push_consumer.go]: done

File: internal/consumer/push_consumer.go (hunk 2 above)
Comment (xujianhai666) [reply internal/consumer/push_consumer.go]: iter can not determine continue or not，recursive is better if user do want to continue to execute

File: primitive/interceptor.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+import (
+	"context"
+)
+
+// PInvoker finish a send invoke on producer.
+type PInvoker func(ctx context.Context, req, reply interface{}) error
+
+// PInterceptor intercepts the execution of a send invoke on producer.
+type PInterceptor func(ctx context.Context, req, reply interface{}, next PInvoker) error
```
Comment (wenfengwang) [inline primitive/interceptor.go]: The PInterceptor and CInterceptor have the same function signature, why not use one definition as below:
```
type Invoker func(ctx context.Context, req, reply interface{}) error

type Interceptor func(ctx context.Context, req, reply interface{}, next Invoker) error
```

File: examples/consumer/simple/main.go [hunk 4]
```diff
@@ -27,14 +27,10 @@ import (
...
-		NameServerAddr: "127.0.0.1:9876",
-		ConsumerModel:  primitive.Clustering,
-		FromWhere:      primitive.ConsumeFromFirstOffset,
-	})
-	err := c.Subscribe("test", primitive.MessageSelector{}, func(ctx *consumer.ConsumeMessageContext,
+	c, _ := consumer.NewPushConsumer("testGroup", "127.0.0.1:9876")
+	err := c.Subscribe("TopicTest", primitive.MessageSelector{}, func(ctx *primitive.ConsumeMessageContext,
 		msgs []*primitive.MessageExt) (primitive.ConsumeResult, error) {
-		fmt.Println(msgs)
+		fmt.Println("subscribe callbacl: %v", msgs)
```
Comment (wenfengwang) [inline examples/consumer/simple/main.go]: callbacl -> callback

File: internal/consumer/push_consumer.go (hunk 2 above)
Comment (wenfengwang) [reply internal/consumer/push_consumer.go]: i mean like this: 
```
interceptor := = interceptors[0]
for i := 1 ; i < len(interceptors); i++ {
  interceptor = func(ctx context.Context, req, reply interface{}, invoker primitive.CInvoker) error {
   return interceptors[i](ctx, req, reply, interceptor)
  }
}
```
so we cloud remove `getChainedInterceptor`

Comment (wenfengwang) [reply internal/consumer/push_consumer.go]: ok

File: primitive/interceptor.go (hunk 3 above)
Comment (wenfengwang) [reply primitive/interceptor.go]: please ignore

File: internal/consumer/push_consumer.go (hunk 1 above)
Comment (wenfengwang) [reply internal/consumer/push_consumer.go]: ok

File: examples/consumer/simple/main.go (hunk 4 above)
Comment (xujianhai666) [reply examples/consumer/simple/main.go]: done

**Comment by wenfengwang** [comment]
LGTM, @ShannonDing @zongtanghu @jonnxu please merge the PR.

**Pull request merged**